src/rust/src/decoder/mod.rs

```diff
@@ -86,10 +88,161 @@ impl dtvcc_ctx {
+                service_number = self.current_packet[pos as usize] & 0x3F; // 6 more significant bits
+                if service_number > 7 {
+                    warn!("dtvcc_process_current_packet: Illegal service number in extended header: {}", service_number);
+                }
+            }
+
+            pos += 1;
+
+            if service_number == 0 && block_length != 0 {
+                // Illegal, but specs say what to do...
+                pos = len; // Move to end
+                break;
+            }
+
+            if block_length != 0 {
+                unsafe {
+                    let mut report = *self.report;
+                    report.services[service_number as usize] = 1;
+                }
+            }
+
+            if service_number > 0 && self.services_active[(service_number - 1) as usize] == 1 {
+                self.process_service_block(service_number - 1, pos, block_length);
+            }
+
+            pos += block_length // Skip data
+        }
+
+        self.clear_packet();
+        if pos != len {
+            // For some reason we didn't parse the whole packet
+            warn!("dtvcc_process_current_packet:  There was a problem with this packet, reseting");
+            self.reset_decoders();
+        }
+
+        if len < 128 && self.current_packet[pos as usize] != 0 {
+            // Null header is mandatory if there is room
+            warn!("dtvcc_process_current_packet: Warning: Null header expected but not found.");
+        }
+    }
+    /// Process service block and call handlers for the respective codesets
+    pub fn process_service_block(&mut self, decoder: u8, pos: u8, block_length: u8) {
+        let mut i = 0;
+        while i < block_length {
+            let curr: usize = (pos + i) as usize;
+
+            let consumed = if self.current_packet[curr] != DTVCC_COMMANDS_C0_CODES_DTVCC_C0_EXT1 {
+                let used = if self.current_packet[curr] <= 0x1F {
+                    self.handle_C0(decoder, pos + i, block_length - i)
+                } else if self.current_packet[curr] <= 0x7F {
+                    self.handle_G0(decoder, pos + i, block_length - i)
+                } else if self.current_packet[curr] <= 0x9F {
+                    self.handle_C1(decoder, pos + i, block_length - i)
+                } else {
+                    self.handle_G1(decoder, pos + i, block_length - i)
+                };
+
+                if used == -1 {
+                    warn!("dtvcc_process_service_block: There was a problem handling the data.");
+                    return;
+                }
+                used
+            } else {
+                let mut used = self.handle_extended_char(decoder, pos + i, block_length - i);
+                used += 1; // Since we had CCX_DTVCC_C0_EXT1
+                used
+            };
+            i += consumed as u8;
+        }
+    }
+    pub fn handle_C0(&mut self, decoder: u8, pos: u8, block_length: u8) -> i32 {
+        unsafe {
```

Comment:
I don't believe that each command processing function needs to be unsafe completely :-) 
Most likely this can be reduced to one single unsafe in the calling function.

Reply:
unsafe is being used here for calling the C function. Once I port the function, unsafe won't be needed